core/src/ast/lifetimes.rs

```diff
@@ -437,3 +394,58 @@ impl Lifetime {
         }
     }
 }
+
+/// Incrementally collect lifetimes that outlive a set of visited lifetimes.
+///
+/// See [`LifetimeEnv::outlives`] for more information.
+pub struct Outlives<'env> {
+    env: &'env LifetimeEnv,
+    visited: Vec<bool>,
+    out: Vec<&'env NamedLifetime>,
+}
+
+impl<'env> Outlives<'env> {
+    /// Returns a new [`Outlives`] based on a given [`LifetimeEnv`].
+    fn new(env: &'env LifetimeEnv) -> Self {
+        Outlives {
+            env,
+            visited: vec![false; env.len()],
+            out: vec![],
+        }
+    }
+
+    /// Visits a `&NamedLifetime`, updating the internal state to mark all lifetimes
+    /// that must outlive the given lifetime as reachable.
+    pub fn visit(&mut self, named: &NamedLifetime) {
+        if let Some(id) = self
+            .env
+            .nodes
+            .iter()
+            .position(|node| node.lifetime == *named)
+        {
+            self.dfs(id);
+        }
+    }
+
+    /// Performs DFS to visit all longer lifetimes.
```

Comment:
nit: doc should say something like "visit all lifetimes recursively longer than `id`" or something

Reply:
also define the id as an index